README
vCard(3)              User Contributed Perl Documentation             vCard(3)



NNAAMMEE
       vCard - read, write, and edit a single vCard

SSYYNNOOPPSSIISS
           use vCard;

           # create the object
           my $vcard = vCard->new;

           # these methods load vCard data
           # (see method documentation for details)
           $vcard->load_file($filename);
           $vcard->load_string($string);
           $vcard->load_hashref($hashref);

           # simple getters/setters
           $vcard->full_name('Bruce Banner, PhD');
           $vcard->title('Research Scientist');
           $vcard->photo('http://example.com/bbanner.gif');

           # complex getters/setters
           $vcard->phones({
               { type => ['work', 'text'], number => '651-290-1234', preferred => 1 },
               { type => ['home'],         number => '651-290-1111' }
           });
           $vcard->email_addresses({
               { type => ['work'], address => 'bbanner@ssh.secret.army.mil' },
               { type => ['home'], address => 'bbanner@timewarner.com'      },
           });

           # these methods output data in vCard format
           my $file   = $vcard->as_file($filename); # writes to $filename
           my $string = $vcard->as_string;          # returns a string

DDEESSCCRRIIPPTTIIOONN
       A vCard is a digital business card.  vCard and vCard::AddressBook
       provide an API for parsing, editing, and creating vCards.

       This module is built on top of Text::vCard.  It provides a more
       intuitive user interface.

       To handle an address book with several vCard entries in it, start with
       vCard::AddressBook and then come back to this module.

       Note that the vCard RFC requires _v_e_r_s_i_o_n_(_) and _f_u_l_l___n_a_m_e_(_).  This
       module does not check or warn if these conditions have not been met.

EENNCCOODDIINNGG AANNDD UUTTFF--88
       See the 'ENCODING AND UTF-8' section of vCard::AddressBook.

MMEETTHHOODDSS
   llooaadd__hhaasshhrreeff(($$hhaasshhrreeff))
       $hashref should look like this:

           full_name    => 'Bruce Banner, PhD',
           given_names  => ['Bruce'],
           family_names => ['Banner'],
           title        => 'Research Scientist',
           photo        => 'http://example.com/bbanner.gif',
           phones       => [
               { type => ['work'], number => '651-290-1234', preferred => 1 },
               { type => ['cell'], number => '651-290-1111' },
           },
           addresses => [
               { type => ['work'], ... },
               { type => ['home'], ... },
           ],
           email_addresses => [
               { type => ['work'], address => 'bbanner@shh.secret.army.mil' },
               { type => ['home'], address => 'bbanner@timewarner.com' },
           ],

       Returns $self in case you feel like chaining.

   llooaadd__ffiillee(($$ffiilleennaammee))
       Returns $self in case you feel like chaining.

   llooaadd__ssttrriinngg(($$ssttrriinngg))
       Returns $self in case you feel like chaining.  This method assumes
       $string is decoded (but not MIME decoded).

   _a_s___s_t_r_i_n_g_(_)
       Returns the vCard as a string.

   aass__ffiillee(($$ffiilleennaammee))
       Write data in vCard format to $filename.

       Dies if not successful.

SSIIMMPPLLEE GGEETTTTEERRSS//SSEETTTTEERRSS
       These methods accept and return strings.

   _v_e_r_s_i_o_n_(_)
       Version number of the vcard.  Defaults to '4.0'

   _f_u_l_l___n_a_m_e_(_)
       A person's entire name as they would like to see it displayed.

   _t_i_t_l_e_(_)
       A person's position or job.

   _p_h_o_t_o_(_)
       This should be a link. Accepts a string or a URI object.  This method
       always returns a URI object.

       TODO: handle binary images using the data uri schema

   _b_i_r_t_h_d_a_y_(_)
   _t_i_m_e_z_o_n_e_(_)
CCOOMMPPLLEEXX GGEETTTTEERRSS//SSEETTTTEERRSS
       These methods accept and return array references rather than simple
       strings.

   _f_a_m_i_l_y___n_a_m_e_s_(_)
       Accepts/returns an arrayref of family names (aka surnames).

   _g_i_v_e_n___n_a_m_e_s_(_)
       Accepts/returns an arrayref.

   _o_t_h_e_r___n_a_m_e_s_(_)
       Accepts/returns an arrayref of names which don't qualify as
       family_names or given_names.

   _h_o_n_o_r_i_f_i_c___p_r_e_f_i_x_e_s_(_)
       Accepts/returns an arrayref.  eg "[ 'Dr.' ]"

   _h_o_n_o_r_i_f_i_c___s_u_f_f_i_x_e_s_(_)
       Accepts/returns an arrayref.  eg "[ 'Jr.', 'MD' ]"

   _p_h_o_n_e_s_(_)
       Accepts/returns an arrayref that looks like:

         [
           { type => ['work'], number => '651-290-1234', preferred => 1 },
           { type => ['cell'], number => '651-290-1111' },
         ]

   _a_d_d_r_e_s_s_e_s_(_)
       Accepts/returns an arrayref that looks like:

         [
           { type => ['work'], street => 'Main St', preferred => 0 },
           { type      => ['home'],
             pobox     => 1234,
             extended  => 'asdf',
             street    => 'Army St',
             city      => 'Desert Base',
             region    => '',
             post_code => '',
             country   => 'USA',
             preferred => 1,
           },
         ]

   _e_m_a_i_l___a_d_d_r_e_s_s_e_s_(_)
       Accepts/returns an arrayref that looks like:

         [
           { type => ['work'], address => 'bbanner@ssh.secret.army.mil' },
           { type => ['home'], address => 'bbanner@timewarner.com', preferred => 1 },
         ]

AAUUTTHHOORR
       Eric Johnson (kablamo), github ~!at!~ iijo dot org

AACCKKNNOOWWLLEEDDGGEEMMEENNTTSS
       Thanks to Foxtons <http://foxtons.co.uk> for making this module
       possible by donating a significant amount of developer time.



perl v5.18.2                      2014-02-14                          vCard(3)


Changes
Revision history for Perl extension Text::vCard.

3.09 Sun 23 Oct 2016 20:24:06 BST
    - Support Orgs (worthmine)

3.08 Sun  9 Oct 2016 21:56:23 BST
    - Enconde/decode the key value pair to UTF-8 solves (jluis)

3.07 Thu 15 Jan 2015 21:44:21 GMT
    - POD fixes (kablamo)

3.06 Thu 12 Jun 2014 19:55:51 BST
    - Fixed Text::vCard::Node->add_types() (kablamo)
    - Fix/guess for tests failing on windows (kablamo)

3.05 Tue 13 May 2014 18:04:30 BST
    - Another win attempt to fix (kablamo)
    - Get travis working! (kablamo)

3.04 Tue 29 Apr 2014 20:21:52
    - Try a fix for Win32 (kablamo)

3.03 Fri 25 Apr 2014 19:23:43
    - Better error msg for vcards w/o proper line endings (kablamo)
    - Bug fix for names. +test (kablamo)

3.02 Fri 14 Feb 18:35:55 2014
    - Try fix Win32, switch File::Slurp to Path::Tiny (kablamo)

3.01 Mon 10 Feb 11:52:00 2014
    - Try fix Win32 (kablamo)

3.0 Tue Jan 21 15:25:40 GMT 2014
    - Major changes!
    - Initial release of vCard and vCard::AddressBook
    - Fix line unwrapping when reading vcards
    - Rewrote Text::vCard::Node->as_string()
    - Rewrote Text::vCard::AddressBook->export()
    - Redo the way that encoding and decoding works
    - Correctly wrap utf8 encoded strings
    - Fix encoding issues with MIME::Base64 and MIME::QuotedPrint
    - Do a better job escaping and unescaping '\', ';', and ','
    - Fix for writing vcards with ORG values
    - Fix to remove extra semi colon when writing vcards
    - Version defaults to 4.0 when using the new vCard library

2.13 Thu 20 Dec 2012 19:34:44 GMT
    - Under the newer combined RFC6350 - every END:VCARD must
      be followed by \r\n (including last one)- reported by ysth
    - Fix for tests, should now work under Perl 5.17.6 - Ron Savage

2.12 Mon 10 Sep 2012 21:09:45 BST
    - perltidy everything
    - Adding base64 encode/decode (doriantaylor)

2.11 Thu 21 May 2012 08:48:46 BST
    - Merge POD cleanup and improvements (szabgab)

2.10 Tue 11 Jan 2011 21:50:02 GMT
    - Merged in quote-printed code (Christian Varga)
    - Cleaned up the tests and code for quote-printed

2.09 Thu 19 Aug 2010 16:08:20 BST
    - Add set_encoding() for export - Stefan

2.08 Tue Jul 13 08:47:58 BST 2010
    - Doc fix (Blair Christensen)

2.07 Wed 28 Jan 2010 22:16:05 GMT
    - Fix bug in get_simple_type() (Philip A. Prindeville)
    - Simplify Makefile.PL and deprecate unused fields (Philip A. Prindeville)

2.06 Sun 24 Jan 2010 10:02:01 GMT
    - Doc typos (Philip A. Prindeville)
    - Add get_simple_type() and import_data() (Philip A. Prindeville)
    - Doc typos (Philip A. Prindeville)

2.05 Fri  8 Jan 2010 22:28:50 GMT
    - Put back changes which were missed out

2.04 Fri  8 Jan 2010 22:00:20 GMT
    - Improvements to Makefile.PL and require 5.6 - thanks cpanservice
    - Switch to git repo
    - Doc patch

2.03 Sun Mar 16 10:10:54 GMT 2008
    - added 'moniker' and marked 'name' as depreciated

2.02 Sat Mar 15 17:15:16 GMT 2008
    - http://rt.cpan.org/Ticket/Display.html?id=34044
      Clean up tests so they pass in perl 5.10
      thanks to Neil Williams

2.01 Sun Sep 2 2007
    - Stop empty TYPE= being set on export when not needed
      thanks to Vadim for the suggestion

2.00 Sat Oct 21 2006
    - export now gives 'TYPE=x,y' instead of the incorrect 'x;y'

1.99 Wed 30th Aug 2006
    - Added 'quoted-printable' fix [rt.cpan.org #17790]
      this requires MIME::QuotedPrint

1.98 Mon Jul 17 2006
    - Test fix

1.97 Sun Jul  2 2006
    - removed test because I don't think Test::More::is_deaply
      works correctly on: perl 5.8.0 on Solaris x86 9

1.96 Sun 3rd July 2005
    - Removed String::ShellQuote - forgot to PREREQ and not using it

1.95 Sun 26th June 2005
    - Made export() work for address book, still missing encoding
      though.
    - Added basic test for export()
    - A few little clean up things

1.94 Sun 24th April 2005
    - Made check for 'VCARD' tag case insensative,
      probably won't help until Text::vFile::asData does as well.

1.93 Mon 31st Jan 2005
    - Restructured object so all nodes are stored in a hashref
      $self->{nodes} so I can get get_group to work, all other
      methods updated to reflect this.
    - added get_group() to vcard

1.92
    - Added group() to Node and extract that info
      from the vcard if it is there.

1.91
    - Doc bug, said 'tels', should have been 'tel'.

1.9
    - removed import() from docs as not used now
    - changed Node so excepts data with less than field order elements
      e.g. N, accepts 'Jon;Smith', As well as 'Jon;Smith;Dr;' this means
      we work better according to the RFC
    - ORGs' 'unit' method requires and returns an array ref as it is
      unlimited in the number of elements it has

1.8 Fri 15th Oct 2004
    - Applied patch from Aaron Cope for version stuff, can
      now access it as lowercase version().

1.7 Wed 13th Oct 2004
    - remembered to update module versions as well as makefile!

1.6 Wed 13th Oct 2004
    - removed import() and made anonsub again
    - removed version() as was conflicting with $VERSION,
      info can be accessed through other methods in anycase

1.5 Wed  Oct 6th 2004
    - Change to support asData change for 'type=home;type=pref'
    - Change to support escaped ; (\;) in data - 'note' now works

1.4 Tue  Oct 5th 2004
    - Patches from Arron for XML::Generator::vCard
    - put auto methods in sub rather than closure for vCard.pm
    - change read_file() to scalar read_file() in address

1.3 Tue Oct 5th 2004
    - Added missing File::Slurp dependency
    - Renamed 'element_type' to 'node_type' as well
    - doc updates with help from binary

1.2 Sat Oct 2nd 2004
    - Changed so can parse group.NODE, the group is ignored for now
    - Added test for apple vcard version 3
    - emailed asData author about small buy with types (home/pref)
    - added autogen methods; $vcard->fn(), $vcard->fullname($value)

1.1 Wed Sept 15th 2003
    - Lots of tests added

lib/Text/vCard/Simple.pm
package Text::vCard::Simple;

use Moo;
use Text::vCard;
use Path::Class;

=head1 SYNOPSIS

    use vCard;

    # create the object
    my $vcard = vCard->new;

    # if you want to parse vcard data from a file see L<vCard::AddressBook>

    # there are 3 ways to load vcard data in one fell swoop 
    # (see method documentation for details)
    $vcard->load_file($filename); 
    $vcard->load_string($string); 
    $vcard->load_hashref($hashref); 

    # there are 3 ways to output data in vcard format
    my $file   = $vcard->as_file($filename); # writes to $filename
    my $string = $vcard->as_string;          # returns a string
    print "$vcard";                          # overloaded as a string

    # simple getters/setters
    $vcard->fullname('Bruce Banner, PhD');
    $vcard->first_name('Bruce');
    $vcard->family_name('Banner');
    $vcard->title('Research Scientist');
    $vcard->photo('http://example.com/bbanner.gif');

    # phones setter example
    $vcard->phones({
        work => { number => '651-290-1234', preferred => 1 },
        cell => { number => '651-290-1111' }
    });

    # phones getter example
    if (my $work_phone = $vcard->phones->work) {
        $work_phone->number();
    }
    $vcard->phones->work->number();
    $vcard->phones->work->preferred();

    # addresses setter example
    $vcard->addresses({
        work => { street => 'Main St' },
        home => { street => 'Army St' },
    });

    # addresses getter example
    $vcard->addresses->work->po_box();
    $vcard->addresses->work->extended();
    $vcard->addresses->work->street();
    $vcard->addresses->work->city();
    $vcard->addresses->work->region();
    $vcard->addresses->work->post_code();
    $vcard->addresses->work->country();
    $vcard->addresses->work->preferred();

    # email setter example
    $vcard->email({
        work => { address => 'bbanner@ssh.secret.army.mil' },
        home => { address => 'bbanner@timewarner.com'      },
    });

    # email getter examples
    $vcard->email->work->address();
    $vcard->email->home->address();
    $vcard->email->home->preferred();


=head1 DESCRIPTION

A vCard is a digital business card.  vCard and vCard::AddressBook provide an
API for parsing, editing, and creating vCards.

This module is built on top of Text::vCard.  It provides a more intuitive user
interface.  

To handle an address book with several vCard entries in it, start with
L<vCard::AddressBook> and then come back to this module.

=head1 METHODS

=cut

has _data => ( is => 'rw' );

=head2 load_hashref($hashref)

$hashref looks like this:

    fullname   => 'Bruce Banner, PhD',
    first_name  => 'Bruce',
    family_name => 'Banner', # required
    title       => 'Research Scientist',
    photo       => 'http://example.com/bbanner.gif',
    phones      => {
        {   type      => 'work',
            number    => '651-290-1234',
            preferred => 1,
        },
        {   type   => 'cell',
            number => '651-290-1111'
        },
    },
    addresses => {
        work => {  },
        home => {  },
    },
    email_addresses => {
        work => { address => 'bbanner@shh.secret.army.mil', preferred => 1 },
        home => { address => 'bbanner@timewarner.com' },
    },

Returns $self in case you feel like chaining.

=cut

sub load_hashref {
    my ( $self, $hashref ) = @_;
    $self->_data($hashref);
    return $self;
}

=head2 as_string()

Returns the vCard as a string.

=cut

sub as_string {
    my ($self) = @_;
    my $vcard = Text::vCard->new;

    my $phones          = $self->_data->{phones};
    my $addresses       = $self->_data->{addresses};
    my $email_addresses = $self->_data->{email_addresses};

    $self->_build_simple_nodes( $vcard, $self->_data );
    $self->_build_phone_nodes( $vcard, $phones ) if $phones;
    $self->_build_address_nodes( $vcard, $addresses ) if $addresses;
    $self->_build_email_address_nodes( $vcard, $email_addresses )
        if $email_addresses;

    return $vcard->as_string;
}

sub _build_simple_nodes {
    my ( $self, $vcard, $data ) = @_;

    my @simple_node_types = qw/fullname title photo birthday timezone/;

    foreach my $node_type (@simple_node_types) {
        next unless $data->{$node_type};
        $vcard->$node_type( $data->{$node_type} );
    }
}

sub _build_phone_nodes {
    my ( $self, $vcard, $phones ) = @_;

    foreach my $phone (@$phones) {

        # TODO: better error handling
        die "'number' attr missing from 'phones'" unless $phone->{number};

        my $type      = $phone->{type};
        my $preferred = $phone->{preferred};
        my $number    = $phone->{number};

        my $params = [];
        push @$params, { type => $type }      if $type;
        push @$params, { pref => $preferred } if $preferred;

        $vcard->add_node(
            {   node_type => 'TEL',
                data      => [ { params => $params, value => $number } ],
            }
        );
    }
}

sub _build_address_nodes {
    my ( $self, $vcard, $addresses ) = @_;

    foreach my $address (@$addresses) {

        my $type      = $address->{type};
        my $preferred = $address->{preferred};

        my $params = [];
        push @$params, { type => $type }      if $type;
        push @$params, { pref => $preferred } if $preferred;

        my $value = join ';',
            $address->{pobox}     || '',
            $address->{extended}  || '',
            $address->{street}    || '',
            $address->{city}      || '',
            $address->{region}    || '',
            $address->{post_code} || '',
            $address->{country}   || '';

        $vcard->add_node(
            {   node_type => 'ADR',
                data      => [ { params => $params, value => $value } ],
            }
        );
    }
}

sub _build_email_address_nodes {
    my ( $self, $vcard, $email_addresses ) = @_;

    foreach my $email_address (@$email_addresses) {

        # TODO: better error handling
        die "'address' attr missing from 'email_addresses'"
            unless $email_address->{address};

        my $type      = $email_address->{type};
        my $preferred = $email_address->{preferred};

        my $params = [];
        push @$params, { type => $type }      if $type;
        push @$params, { pref => $preferred } if $preferred;

        # TODO: better error handling
        my $value = $email_address->{address};

        $vcard->add_node(
            {   node_type => 'EMAIL',
                data      => [ { params => $params, value => $value } ],
            }
        );
    }
}

=head2 as_file($filename)

Write data in vCard format to $filename.

Returns a Path::Class::File if successful.  Dies if not successful.

=cut

sub as_file {
    my ( $self, $filename ) = @_;

    my $file = ref $filename eq 'Path::Class::File'    #
        ? $filename
        : file($filename);

    $file->spew( $self->as_string );

    return $file;
}

1;


lib/vCard/Role/FileIO.pm
package vCard::Role::FileIO;

use Moo::Role;
use Path::Tiny;

requires qw/encoding_in encoding_out/;

# PerlIO layers should look like ':encoding(UTF-8)'
# The ':encoding()' part does character set and encoding transformations.
# Without it you are just declaring the stream to be of a certain encoding.
# See PerlIO, PerlIO::encoding docs.

sub _iomode_out {
    my ($self) = @_;
    return { binmode => ':raw' } if $self->encoding_out eq 'none';
    return { binmode => ':raw:encoding(' . $self->encoding_out . ')' };
}

sub _iomode_in {
    my ($self) = @_;
    return { binmode => ':raw' } if $self->encoding_in eq 'none';
    return { binmode => ':raw:encoding(' . $self->encoding_in . ')' };
}

# Filename can be a string, a Path::Tiny obj, or a Path::Class obj.
# Returns a Path::Tiny obj.
sub _path {
    my ( $self, $filename ) = @_;
    return ref $filename eq 'Path::Class::File'    #
        ? path("$filename")
        : path($filename);    # works for strings and Path::Tiny objects
}

1;


lib/vCard.pm
package vCard;

use Moo;

use Carp;
use Path::Tiny;
use Text::vCard;
use vCard::AddressBook;
use URI;

=head1 NAME

vCard - Read, write, and edit vCards

=head1 SYNOPSIS

    use vCard;

    # create the object
    my $vcard = vCard->new;

    # these methods load vCard data
    # (see method documentation for details)
    $vcard->load_file($filename); 
    $vcard->load_string($string); 
    $vcard->load_hashref($hashref); 

    # simple getters/setters
    $vcard->full_name('Bruce Banner, PhD');
    $vcard->title('Research Scientist');
    $vcard->photo('http://example.com/bbanner.gif');

    # complex getters/setters
    $vcard->phones([
        { type => ['work', 'text'], number => '651-290-1234', preferred => 1 },
        { type => ['home'],         number => '651-290-1111' }
    ]);
    $vcard->email_addresses([
        { type => ['work'], address => 'bbanner@ssh.secret.army.mil' },
        { type => ['home'], address => 'bbanner@timewarner.com'      },
    ]);

    # these methods output data in vCard format
    my $file   = $vcard->as_file($filename); # writes to $filename
    my $string = $vcard->as_string;          # returns a string


=head1 DESCRIPTION

A vCard is a digital business card.  vCard and L<vCard::AddressBook> provide an
API for parsing, editing, and creating vCards.

This module is built on top of L<Text::vCard>.  It provides a more intuitive user
interface.  

To handle an address book with several vCard entries in it, start with
L<vCard::AddressBook> and then come back to this module.

Note that the vCard RFC requires version() and full_name().  This module does
not check or warn if these conditions have not been met.


=head1 ENCODING AND UTF-8

See the 'ENCODING AND UTF-8' section of L<vCard::AddressBook>.


=head1 METHODS

=cut

has encoding_in  => ( is => 'rw', default => sub {'UTF-8'} );
has encoding_out => ( is => 'rw', default => sub {'UTF-8'} );
has _data        => ( is => 'rw', default => sub { { version => '4.0' } } );

with 'vCard::Role::FileIO';

=head2 load_hashref($hashref)

$hashref should look like this:

    full_name    => 'Bruce Banner, PhD',
    given_names  => ['Bruce'],
    family_names => ['Banner'],
    title        => 'Research Scientist',
    photo        => 'http://example.com/bbanner.gif',
    phones       => [
        { type => ['work'], number => '651-290-1234', preferred => 1 },
        { type => ['cell'], number => '651-290-1111' },
    },
    addresses => [
        { type => ['work'], ... },
        { type => ['home'], ... },
    ],
    email_addresses => [
        { type => ['work'], address => 'bbanner@shh.secret.army.mil' },
        { type => ['home'], address => 'bbanner@timewarner.com' },
    ],

Returns $self in case you feel like chaining.

=cut

sub load_hashref {
    my ( $self, $hashref ) = @_;
    $self->_data($hashref);

    $self->_data->{version} = '4.0'
        unless $self->_data->{version};

    $self->_data->{photo} = URI->new( $self->_data->{photo} )
        unless ref $self->_data->{photo} =~ /^URI/;

    return $self;
}

=head2 load_file($filename)

Returns $self in case you feel like chaining.

=cut

sub load_file {
    my ( $self, $filename ) = @_;

    my $addressBook = vCard::AddressBook->new({
        encoding_in  => $self->encoding_in,
        encoding_out => $self->encoding_out,
    });
    my $vcard = $addressBook->load_file($filename)->vcards->[0];

    $self->_data($vcard->_data);

    return $self;
}

=head2 load_string($string)

Returns $self in case you feel like chaining.  This method assumes $string is
decoded (but not MIME decoded).

=cut

sub load_string {
    my ( $self, $string ) = @_;

    my $addressBook = vCard::AddressBook->new({
        encoding_in  => $self->encoding_in,
        encoding_out => $self->encoding_out,
    });
    my $vcard = $addressBook->load_string($string)->vcards->[0];

    $self->_data($vcard->_data);

    return $self;
}

=head2 as_string()

Returns the vCard as a string.

=cut

sub as_string {
    my ($self) = @_;
    my $vcard = Text::vCard->new( { encoding_out => $self->encoding_out } );

    my $phones          = $self->_data->{phones};
    my $addresses       = $self->_data->{addresses};
    my $email_addresses = $self->_data->{email_addresses};

    $self->_build_simple_nodes( $vcard, $self->_data );
    $self->_build_name_node( $vcard, $self->_data );
    $self->_build_org_node( $vcard, $self->_data->{org}  ) if $self->_data->{org};
    $self->_build_phone_nodes( $vcard, $phones ) if $phones;
    $self->_build_address_nodes( $vcard, $addresses ) if $addresses;
    $self->_build_email_address_nodes( $vcard, $email_addresses )
        if $email_addresses;

    return $vcard->as_string;
}

sub _simple_node_types {
    qw/full_name title photo birthday timezone version/;
    #geo, too?
}

sub _build_simple_nodes {
    my ( $self, $vcard, $data ) = @_;

    foreach my $node_type ( $self->_simple_node_types ) {
        if ( $node_type eq 'full_name' ) {
            next unless $data->{full_name};
            $vcard->fullname( $data->{full_name} );
        } else {
            next unless $data->{$node_type};
            $vcard->$node_type( $data->{$node_type} );
        }
    }
}

sub _build_complex_node {
    my ( $self, $vcard, $node_type, $data ) = @_;
    croak '$data must be HASHREF' unless ref $data eq 'HASH';
    $vcard->add_node( { node_type => $node_type, data => [ $data ] } );
}

sub _build_org_node {
    my ( $self, $vcard, $data ) = @_;
    
    my $value = join ';', @{ $data || [] };
    $self->_build_complex_node( $vcard, 'ORG', { value => $value } );
}

sub _build_name_node {
    my ( $self, $vcard, $data ) = @_;

    my $value = join ',', @{ $data->{family_names} || [] };
    $value .= ';' . join ',', @{ $data->{given_names}        || [] };
    $value .= ';' . join ',', @{ $data->{other_names}        || [] };
    $value .= ';' . join ',', @{ $data->{honorific_prefixes} || [] };
    $value .= ';' . join ',', @{ $data->{honorific_suffixes} || [] };


    $self->_build_complex_node( $vcard, 'N', { value => $value } )
        if $value ne ';;;;';
}

sub _build_phone_nodes {
    my ( $self, $vcard, $phones ) = @_;

    foreach my $phone (@$phones) {

        # TODO: better error handling
        croak "'number' attr missing from 'phones'" unless $phone->{number};
        croak "'type' attr in 'phones' should be an arrayref"
            if ( $phone->{type} && ref( $phone->{type} ) ne 'ARRAY' );

        my $type      = $phone->{type} || [];
        my $preferred = $phone->{preferred};
        my $number    = $phone->{number};

        my $params = [];
        push @$params, { type => $_ } foreach @$type;
        push @$params, { pref => $preferred } if $preferred;

        $self->_build_complex_node( $vcard, 'TEL', { params => $params, value => $number } );
    }
}

sub _build_address_nodes {
    my ( $self, $vcard, $addresses ) = @_;

    foreach my $address (@$addresses) {

        croak "'type' attr in 'addresses' should be an arrayref"
            if ( $address->{type} && ref( $address->{type} ) ne 'ARRAY' );

        my $type = $address->{type} || [];
        my $preferred = $address->{preferred};

        my $params = [];
        push @$params, { type => $_ } foreach @$type;
        push @$params, { pref => $preferred } if $preferred;

        my $value = join ';',
            $address->{pobox}     || '',
            $address->{extended}  || '',
            $address->{street}    || '',
            $address->{city}      || '',
            $address->{region}    || '',
            $address->{post_code} || '',
            $address->{country}   || '';

        $self->_build_complex_node( $vcard, 'ADR', { params => $params, value => $value } );
    }
}

sub _build_email_address_nodes {
    my ( $self, $vcard, $email_addresses ) = @_;

    foreach my $email_address (@$email_addresses) {

        # TODO: better error handling
        croak "'address' attr missing from 'email_addresses'"
            unless $email_address->{address};
        croak "'type' attr in 'email_addresses' should be an arrayref"
            if ( $email_address->{type}
            && ref( $email_address->{type} ) ne 'ARRAY' );

        my $type = $email_address->{type} || [];
        my $preferred = $email_address->{preferred};

        my $params = [];
        push @$params, { type => $_ } foreach @$type;
        push @$params, { pref => $preferred } if $preferred;

        # TODO: better error handling
        my $value = $email_address->{address};

        $self->_build_complex_node( $vcard, 'EMAIL', { params => $params, value => $value } );
    }
}

=head2 as_file($filename)

Write data in vCard format to $filename.

Dies if not successful.

=cut

sub as_file {
    my ( $self, $filename ) = @_;
    my $file = $self->_path($filename);
    $file->spew( $self->_iomode_out, $self->as_string );
    return $file;
}

=head1 SIMPLE GETTERS/SETTERS

These methods accept and return strings.  

=head2 version()

Version number of the vcard.  Defaults to '4.0'

=head2 full_name()

A person's entire name as they would like to see it displayed.  

=head2 title()

A person's position or job.

=head2 photo()

This should be a link. Accepts a string or a URI object.  This method
always returns a L<URI> object. 

TODO: handle binary images using the data uri schema

=head2 birthday()

=head2 timezone()


=head1 COMPLEX GETTERS/SETTERS

These methods accept and return array references rather than simple strings.

=head2 family_names()

Accepts/returns an arrayref of family names (aka surnames).

=head2 given_names()

Accepts/returns an arrayref.

=head2 other_names()

Accepts/returns an arrayref of names which don't qualify as family_names or
given_names.

=head2 honorific_prefixes()

Accepts/returns an arrayref.  eg C<[ 'Dr.' ]>

=head2 honorific_suffixes()

Accepts/returns an arrayref.  eg C<[ 'Jr.', 'MD' ]>

=head2 phones()

Accepts/returns an arrayref that looks like:

  [
    { type => ['work'], number => '651-290-1234', preferred => 1 },
    { type => ['cell'], number => '651-290-1111' },
  ]

=head2 addresses()

Accepts/returns an arrayref that looks like:

  [
    { type => ['work'], street => 'Main St', preferred => 0 },
    { type      => ['home'], 
      pobox     => 1234,
      extended  => 'asdf',
      street    => 'Army St',
      city      => 'Desert Base',
      region    => '',
      post_code => '',
      country   => 'USA',
      preferred => 1,
    },
  ]

=head2 email_addresses()

Accepts/returns an arrayref that looks like:

  [
    { type => ['work'], address => 'bbanner@ssh.secret.army.mil' },
    { type => ['home'], address => 'bbanner@timewarner.com', preferred => 1 },
  ]

=cut

sub version             { shift->_setget( 'version',            @_ ) }
sub full_name           { shift->_setget( 'full_name',          @_ ) }
sub family_names        { shift->_setget( 'family_names',       @_ ) }
sub given_names         { shift->_setget( 'given_names',        @_ ) }
sub other_names         { shift->_setget( 'other_names',        @_ ) }
sub honorific_prefixes  { shift->_setget( 'honorific_prefixes', @_ ) }
sub honorific_suffixes  { shift->_setget( 'honorific_suffixes', @_ ) }
sub title               { shift->_setget( 'title',              @_ ) }
sub photo               { shift->_setget( 'photo',              @_ ) }
sub birthday            { shift->_setget( 'birthday',           @_ ) }
sub timezone            { shift->_setget( 'timezone',           @_ ) }
sub phones              { shift->_setget( 'phones',             @_ ) }
sub addresses           { shift->_setget( 'addresses',          @_ ) }
sub email_addresses     { shift->_setget( 'email_addresses',    @_ ) }
sub organization        { shift->_setget( 'organization',       @_ ) }

sub _setget {
    my ( $self, $attr, $value ) = @_;

    $value = URI->new($value)
        if $value && $attr eq 'photo' && ref $value =~ /^URI/;

    $self->_data->{$attr} = $value if $value;

    return $self->_data->{$attr};
}

=head1 AUTHOR

Eric Johnson (kablamo), github ~!at!~ iijo dot org

=head1 ACKNOWLEDGEMENTS

Thanks to L<Foxtons|http://foxtons.co.uk> for making this module possible by
donating a significant amount of developer time.

=cut

1;
